### PR TITLE
Importprocessor

### PR DESCRIPTION
**Pull request by TimsLuxo**
Added IImportPostProcessHierarchy and IImportPostProcessGeometry interfaces that are run by UsdAsset at import time to allow user modification of the incoming scene. Two working example implementation. IImportProcessUsd stubbed in, not implemented.

### CODE REVIEW

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs [hunk 1]
```diff
@@ -338,9 +321,10 @@ public ImportException(string message, System.Exception innerException)
 
       if (ShouldYield(targetTime, timer)) { yield return null; ResetTimer(timer); }
 
-      Profiler.BeginSample("USD: AfterBuildPrimMap");
-      if (AfterBuildPrimMap != null) {
-        AfterBuildPrimMap(primMap, scene, usdPrimRoot, root, importOptions);
+      Profiler.BeginSample("USD: PostProcessHierarchy");
+      foreach( var processor in root.GetComponents<IImportPostProcessHierarchy>())
+      {
+        processor.PostProcessHierarchy(primMap, importOptions);
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs]: We should wrap all callbacks in a try catch, like this:

```
try {
   processor.PostProcessHierarchy(primMap, importOptions);
} catch (Exception ex) {
   Debug.LogException(ex);
}
```

And I just noticed the old code wasn't doing this, but it really should have been.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs [hunk 2]
```diff
@@ -929,12 +913,10 @@ public ImportException(string message, System.Exception innerException)
...
-      //
-      // AfterImport callback.
-      //
-      Profiler.BeginSample("AfterImport callback");
-      if (AfterImport != null) {
-        AfterImport(primMap, scene, usdPrimRoot, root, importOptions);
+      Profiler.BeginSample("USD: PostProcessGeometry");
+      foreach( var processor in root.GetComponents<IImportPostProcessGeometry>())
+      {
+        processor.PostProcessGeometry(primMap, importOptions);
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs]: Add try catch (see comment above)

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+﻿using System.Collections;
+using System.Collections.Generic;
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs]: Add license notice

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿using System.Collections;
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs]: Add license

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿using System.Collections;
```
Comment (jcowles) [inline package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs]: Add license

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+using UnityEngine;
+using System.Text.RegularExpressions;
+using pxr;
+
+//possibly move this to Examples
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs]: Done, moved into Unity.Formats.USD.Examples namespace, Samples~/ directory.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
+﻿using System.Collections;
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs]: License added to all .cs files

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using USD.NET;
+using pxr;
+namespace Unity.Formats.USD {
+
+    // happens while hierarchy is loading
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs]: Doc-style comments added and Tooltips added

**Comment by TimsLuxo** [comment]
Comments addressed, also renamed IImportPostProcessGeometry to IImportPostProcessComponents as it can operate on any component (camera, etc.)

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    public interface IImportProcessUsd
+    {
+        void ProcessUsd(Scene scene, SceneImportOptions sceneImportOptions);
+    }
+
+    // happens after GameObject Hierarchy created
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs]: Doc comment added.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/RegexImportProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using System.Text.RegularExpressions;
+using pxr;
+
+//possibly move this to Examples
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/RegexImportProcessor.cs]: moved...

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+using System.Text.RegularExpressions;
+using pxr;
+
+//possibly move this to Examples
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs]: moved...

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+namespace Unity.Formats.USD
+{
+    public class SetHideFlags : RegexImportProcessor, IImportPostProcessHierarchy
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs]: doc comment added

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/RegexImportProcessor.cs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+namespace Unity.Formats.USD
+{
+    public abstract class RegexImportProcessor : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/RegexImportProcessor.cs]: doc comment added

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs (hunk 3 above)
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/CombineMeshes.cs]: license added

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs (hunk 4 above)
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs]: license added

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs (hunk 5 above)
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/SetHideFlags.cs]: license added

File: package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+    public interface IImportPostProcessHierarchy
+    {
+        void PostProcessHierarchy(PrimMap primMap, SceneImportOptions sceneImportOptions);
+    }
+
+    // happens after GameObject Hierarchy created and Geometry created
```
[the remark this answers is not in the document]
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/ImportProcessor/ImportProcessor.cs]: doc comment added

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs (hunk 1 above)
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs]: ok, wrapped.

File: package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs (hunk 2 above)
Comment (TimsLuxo) [reply package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneImporter.cs]: ok, wrapped.

**Pull request merged**